### PR TITLE
[10.0][ADD] Add secure response for base_rest module

### PR DESCRIPTION
**Pull request by acsonefho**
- Add a validation (same as input) for the json response
- Add return validator to avoid failure on existing unit-tests (could be adapted later)

### CODE REVIEW

File: base_rest/components/service.py [hunk 1]
```diff
@@ -47,6 +47,15 @@ def skip_secure_params(func):
     func.skip_secure_params = True
 
 
+def skip_secure_response(func):
```
Comment (lmignon) [inline base_rest/components/service.py]: @acsonefho You can remove the skip decorators.

File: base_rest/components/service.py [hunk 2]
```diff
@@ -102,6 +117,29 @@ def _secure_params(self, method, params):
...
+        :param response: dict/json
+        :return: dict/json
+        """
+        method_name = method
+        if callable(method):
+            method_name = method.__name__
+        if hasattr(method, 'skip_secure_response'):
+            return response
+        schema = self._get_schema_output_for_method(method_name)
+        v = Validator(schema, purge_unknown=True)
```
Comment (lmignon) [inline base_rest/components/service.py]: @acsonefho You can remove 'purge_unknown' for the response. We must be sure that the provided informations is conform to what is expected and in the same time avoid to lost time searching why a new field is not present into the response if the schema has not been adapted

File: base_rest/components/service.py [hunk 3]
```diff
@@ -80,6 +89,12 @@ def _get_schema_for_method(self, method_name):
             raise NotImplementedError(validator_method)
         return getattr(self, validator_method)()
 
+    def _get_schema_output_for_method(self, method_name):
+        validator_method = '_validator_return_%s' % method_name
```
Comment (lmignon) [inline base_rest/components/service.py]: @acsonefho IMO we should change the naming convention for methods in change of providing the schemas. 
May be '_my_method_request_schema'  '_my_method_response_schema'. In the same time we should replace these methods by properties. `@property` @sebastienbeau What do you think about?

**Review by lmignon — CHANGES_REQUESTED** [review]

File: base_rest/components/service.py (hunk 1 above)
Comment (acsonefho) [reply base_rest/components/service.py]: @lmignon The existing 'skip_secure_params' too?

Comment (lmignon) [reply base_rest/components/service.py]: @acsonefho decorator**s** -> plural means also 'skip_secure_params' :smirk:

File: base_rest/README.rst [hunk 4]
```diff
@@ -50,19 +50,29 @@ The business logic of your service must be implemented into a component
...
 
-        # Validator
-        def _validator_search(self):
+        # Validator (input)
+        def _search_request_schema(self):
             return {'message': {'type': 'string'}}
 
         # Validator
-        def _validator_get(self):
+        def _get_request_schema(self):
```
Comment (lmignon) [inline base_rest/README.rst]: @acsonefho `@property`
Can you also group properties by method? `_get_request_schema,  _get_response_schema`...

File: base_rest/README.rst [hunk 5]
```diff
@@ -50,19 +50,29 @@ The business logic of your service must be implemented into a component
...
             # no parameters by default
             return {}
 
-        def _validator_update(self):
+        def _update_request_schema(self):
```
Comment (lmignon) [inline base_rest/README.rst]: @acsonefho `@property`

File: base_rest/README.rst [hunk 6]
```diff
@@ -50,19 +50,29 @@ The business logic of your service must be implemented into a component
...
+            return {'message': {'type': 'string'}}
+
+        def _create_request_schema(self):
```
Comment (lmignon) [inline base_rest/README.rst]: @acsonefho `@property`

File: base_rest/README.rst [hunk 7]
```diff
@@ -50,19 +50,29 @@ The business logic of your service must be implemented into a component
...
+            return {'message': {'type': 'string'}}
+
+        # Response Validator (output)
```
Comment (lmignon) [inline base_rest/README.rst]: @acsonefho `@property`

File: base_rest/README.rst [hunk 8]
```diff
@@ -50,19 +50,29 @@ The business logic of your service must be implemented into a component
...
+        def _get_response_schema(self):
+            return {'id': {'type': 'integer'}}
+
+        def _update_response_schema(self):
```
Comment (lmignon) [inline base_rest/README.rst]: @acsonefho `@property`

File: base_rest/README.rst [hunk 9]
```diff
@@ -50,19 +50,29 @@ The business logic of your service must be implemented into a component
...
             return {'message': {'type': 'string'}}
 
-        def _validator_create(self):
+        def _create_response_schema(self):
```
Comment (lmignon) [inline base_rest/README.rst]: @acsonefho `@property`

File: base_rest/components/service.py [hunk 10]
```diff
@@ -84,32 +66,28 @@ def _log_call(self, func, params, secure_params, res):
             _logger.debug(message, *args, extra=extra)
 
     def _get_schema_for_method(self, method_name):
```
Comment (lmignon) [inline base_rest/components/service.py]: @acsonefho `get_request_schema(self, method_name)` to be consistent with the naming of schema properties

File: base_rest/components/service.py [hunk 11]
```diff
@@ -84,32 +66,28 @@ def _log_call(self, func, params, secure_params, res):
...
-        validator_method = '_validator_%s' % method_name
+        validator_method = '_%s_request_schema' % method_name
         if not hasattr(self, validator_method):
             raise NotImplementedError(validator_method)
-        return getattr(self, validator_method)()
+        return getattr(self, validator_method)
 
     def _get_schema_output_for_method(self, method_name):
-        validator_method = '_validator_return_%s' % method_name
```
Comment (lmignon) [inline base_rest/components/service.py]: @acsonefho `get_response_schema(self, method_name)` to be consistent with the naming of schema properties

**Review by lmignon — CHANGES_REQUESTED** [review]

File: base_rest/components/service.py [hunk 12]
```diff
@@ -75,33 +66,54 @@ def _log_call(self, func, params, secure_params, res):
             _logger.debug(message, *args, extra=extra)
 
     def _get_schema_for_method(self, method_name):
-        validator_method = '_validator_%s' % method_name
+        validator_method = '_%s_request_schema' % method_name
+        if not hasattr(self, validator_method):
+            raise NotImplementedError(validator_method)
+        return getattr(self, validator_method)
+
+    def _get_schema_output_for_method(self, method_name):
```
Comment (lmignon) [inline base_rest/components/service.py]: @acsonefho `get_response_schema(self, method_name)` to be consistent with the naming of schema properties

File: base_rest/components/service.py (hunk 3 above)
Comment (lmignon) [reply base_rest/components/service.py]: @sebastienbeau  @sbidoul Do you have some advice on the naming convention to apply to the properties providing the schema to validate the request and the response?
`_(method_name)_(request|response)_schema`

```python
@property
_get_request_schema(self):
    pass

@property
_get_response_schema(self):
    pass

@property
_search_request_schema(self):
    pass

@property
_search_response_schema(self):
    pass

@property
_update_request_schema(self):
    pass

@property
_update_response_schema(self):
    pass
```
or
`_schema_(method_name)_(request|response)`

```python
@property
_schema_get_request(self):
    pass

@property
_schema_get_response(self):
    pass

@property
_schema_search_request(self):
    pass

@property
_schema_search_response(self):
    pass

@property
_schema_update_request(self):
    pass

@property
_schema_update_response(self):
    pass
```
or
???

**Comment by acsonefho** [comment]
@lmignon Should I create a response validator for every Shopinvader module?
I did for shopinvader and shopinvader_payment_stripe but still have errors on shopinvader_delivery_carrier, shopinvader_payment, shopinvader_image,...

**Comment by lmignon** [comment]
@acsonefho yes you should

**Comment by sebastienbeau** [comment]
@lmignon as we plan to do two refactor
- moving specific code (store_cache, erp_session) to shopinvader_locomotive
- refactor the way to declare the validator (using class instead of simple dict)
I think it's simplier to do that only on the next version? I fear that for this version it will request too much time to maintain cerberus schema? What do you think?

**Comment by lmignon** [comment]
@sebastienbeau It's a hard job. We have no resources for this job for now.

**Comment by sebastienbeau** [comment]
Yes, so for now I close the PR and we will see later with the new API

**Pull request closed without merging**